### PR TITLE
Move polarity into `PolyTraitRef` rather than storing it on the side

### PR DESCRIPTION
**Pull request by compiler-errors**
Arguably we could move these modifiers into `TraitRef` instead of `PolyTraitRef`, but I see `TraitRef` as simply the *path* part of the trait ref. It doesn't really matter -- refactoring this further is much easier now.

### CODE REVIEW

**Review by Nadrieril — APPROVED** [review]
Looks reasonable and like a nice improvement! I lack all context particularly for the first commit though so I'll let someone more knowledgeable call this.

r? compiler

**Comment by chenyukang** [comment]
r? compiler

**Review by jieyouxu — APPROVED** [review]
Looks reasonable to me as well. As to removing the hack, if no tests fail then I think we're good 🤷 You can r=Nadrieril and me after a rebase.

File: compiler/rustc_hir_pretty/src/lib.rs [hunk 1]
```diff
@@ -679,6 +676,10 @@ impl<'a> State<'a> {
     }
 
     fn print_poly_trait_ref(&mut self, t: &hir::PolyTraitRef<'_>) {
+        // FIXME: This isn't correct!
+        if t.modifiers == TraitBoundModifier::Maybe {
+            self.word("?");
+        }
```
Comment (jieyouxu) [inline compiler/rustc_hir_pretty/src/lib.rs:682]: Question (not-blocking): I'm guessing these are for follow-ups?

Comment (compiler-errors) [reply compiler/rustc_hir_pretty/src/lib.rs:682]: Yeah, I didn't really want to make any changes to the behavior here. And all of these bugs are on the diagnostic path anyways.

**Pull request merged**